### PR TITLE
Fix missing is_valid() parameter

### PR DESCRIPTION
**Pull request by ksamuel**

### CODE REVIEW

**Comment by ksamuel** [comment]
Hey guys, I noticed the sentinel failing sometime so I found this. 

I'm really interested in making the sentinel more stable because we get way more WATCHDOG_EXPIRED than we should.

I've got 12 years of Python coding behind me, so the code is pretty clear, but I don't know how the system work so if somebody from your team want to team up with me on this, that would be nice.

File: lib/models.py [hunk 1]
```diff
@@ -142,7 +142,7 @@ def import_gobject_from_dashd(self, dashd, rec):
         try:
             newdikt = subdikt.copy()
             newdikt['object_hash'] = object_hash
-            if subclass(**newdikt).is_valid() is False:
+            if subclass(**newdikt).is_valid(dashd) is False:
```
Comment (nmarley) [inline lib/models.py:145]: The classes here (`subclass`) are `Proposal` and `Superblock`, which don't accept any arguments to `is_valid`, so this change doesn't make sense and also isn't valid.

**Comment by nmarley** [comment]
Hi, thanks for contributing! Please see the inline comment above ^. Basically, this change isn't valid and will break the code.

**Comment by ksamuel** [comment]
It's weird, cause I swear I saw errors in the sentinel because of it after I activated `SENTINAL_DEBUG=1`.  I did a little `pdb` into it and passing `dashd` seemed to resolve the problem. 

Are you sure than since: `class Watchdog(BaseModel, GovernanceClass)`, ` subclass(**newdikt)` doesn't pick it up ? Cause `Watchdog.is_valid` does need a `dashd` object.

I didn't read the full code, so I may be missing bits here and there, but there is definitely something going on around this call.

**Comment by ksamuel** [comment]
@nmarley: I bypassed the scheduler (BTW: what are the pros and cons of doing that ? I really don't understand what's is the sentinel protocol) so I could give you more info.

Without the PR, here is what I get:
```
...
('UPDATE "governance_objects" SET "object_hash" = ?, "object_type" = ?, "object_fee_tx" = ?, "yes_count" = ?, "no_count" = ?, "abstain_count" = ?, "absolute_yes_count" = ? WHERE ("governance_objects"."id" = ?)', [u'f7d930c8.....0531eea64f7c0', 3, u'0000000000....00000000000000', 0, 0, 0, 0, 2])
2017-12-28 11:54:27 govobj updated = 1
2017-12-28 11:54:27 Got invalid object from dashd! is_valid() takes exactly 2 arguments (1 given)
('SELECT Count(*) FROM "votes" AS t1 WHERE ((("t1"."governance_object_id" = ?) AND ("t1"."signal_id" = ?)) AND ("t1"."outcome_id" = ?))', [2, 3, 1])
('SELECT Count(*) FROM "votes" AS t1 WHERE (("t1"."governance_object_id" = ?) AND ("t1"."signal_id" = ?))', [2, 3])
...
```

We can clearly see there is a `Got invalid object from dashd! is_valid() takes exactly 2 arguments (1 given)` in there. And it doesn't happen if I pass `dashd`.

The piece of code is:

```
       try:
            newdikt = subdikt.copy()
            newdikt['object_hash'] = object_hash
            if subclass(**newdikt).is_valid() is False:
                govobj.vote_delete(dashd)
                return (govobj, None)

            subobj, created = subclass.get_or_create(object_hash=object_hash, defaults=subdikt)
        except Exception as e:
            # in this case, vote as delete, and log the vote in the DB
            printdbg("Got invalid object from dashd! %s" % e)
            govobj.vote_delete(dashd)
            return (govobj, None)
```
So we get a `TypeError('is_valid() takes exactly 2 arguments (1 given)',) from `if subclass(**newdikt).is_valid() is False:` and the debugger reveals it's a watchdog object:

```
ipdb> subclass(**newdikt)
<models.Watchdog object at 0x7f2daa34dd10>
```

So an alternative code would be:

```
       try:
            newdikt = subdikt.copy()
            newdikt['object_hash'] = object_hash

            sub, params = subclass(**newdikt), []
            if isinstance(sub, Watchdog):
                params = [dashd]

            if sub.is_valid(*params) is False:
                govobj.vote_delete(dashd)
                return (govobj, None)

            subobj, created = subclass.get_or_create(object_hash=object_hash, defaults=subdikt)

        except Exception as e:
            # in this case, vote as delete, and log the vote in the DB
            printdbg("Got invalid object from dashd! %s" % e)
            govobj.vote_delete(dashd)
            return (govobj, None)
```

But to be honest, it's a bit hacky.

It's better to fix the entire interface and always either pass dashd, or never pass dashd. Or make 2 different methods, both called. 

Also, a generic `except Exception` is bound to generate bug like this. It should be replace with the specific exception we except. What's an ` invalid object from dashd` ? Let's define that and make the block more granular.

**Comment by ksamuel** [comment]
@nmarley : should I make it an issue instead of a PR ? Or just move that info to #30 ?

**Comment by nmarley** [comment]
I think you might have a point, but need more time to read thru the code to be sure. Regardless, #34 should remove the need for this anyway and there will only be the one interface for `is_valid`.

**Comment by ksamuel** [comment]
Great ! Is there a doc somewhere that I can read to understand the sentinel voting system ? I have many "watchdog expired" and "new_start_required" and I would like to figure out why, and who to prevent this.

**Pull request closed without merging**